src/modes/battle/battle_finish.cpp

```diff
@@ -441,32 +448,38 @@ void FinishVictoryAssistant::Initialize()
     }
 
     // Collect the XP, drunes, and dropped objects for each defeated enemy
-    std::deque<BattleEnemy *>& all_enemies = BattleMode::CurrentInstance()->GetEnemyActors();
-    std::vector<GlobalObject *> objects;
-    std::map<GlobalObject *, int32_t>::iterator iter;
-
-    for(uint32_t i = 0; i < all_enemies.size(); ++i) {
+    std::deque<BattleEnemy*>& all_enemies = BattleMode::CurrentInstance()->GetEnemyActors();
+    for (uint32_t i = 0; i < all_enemies.size(); ++i) {
         GlobalEnemy* enemy = all_enemies[i]->GetGlobalEnemy();
+        assert(enemy != nullptr);
+
         _xp_earned += enemy->GetExperiencePoints();
         _drunes_dropped += enemy->GetDrunesDropped();
-        enemy->DetermineDroppedObjects(objects);
 
-        for(uint32_t j = 0; j < objects.size(); ++j) {
+        std::vector<GlobalObject*> objects = enemy->DetermineDroppedObjects();
+        for (uint32_t j = 0; j < objects.size(); ++j) {
             // Check if the object to add is already in our list. If so, just increase the quantity of that object.
             // iter = _objects_dropped.find(objects[j]); // Will not work since each item is created with new.
             // Need to search for the item ID instead.
-            iter = _objects_dropped.begin();
-            while(iter != _objects_dropped.end()) {
-                if(iter->first->GetID() == objects[j]->GetID()) break;
+            auto iter = _objects_dropped.begin();
+            while (iter != _objects_dropped.end()) {
+                if (iter->first->GetID() == objects[j]->GetID()) {
+                    break;
+                }
                 ++iter;
             }
 
-            if(iter != _objects_dropped.end()) {
+            if (iter != _objects_dropped.end()) {
                 iter->second++;
+
+                // Clean up the global object.
+                delete objects[j];
```

Comment:
Here's a minor clean up if the object is a duplicate.  For example, if the enemies drop two health potions in a single battle.